Template components (take 2) (#7425)
wagtail/wagtail: docs/extending/template_components.md

# Template components

> Working with objects that know how to render themselves as elements on an HTML template is a common pattern seen throughout the Wagtail admin. For example, the admin homepage is a view provided by the central `wagtail.admin` app, but brings together information panels sourced from various other modules of Wagtail, such as images and documents (potentially along with others provided by third-party packages). These panels are passed to the homepage via the [`construct_homepage_panels`](../reference/hooks.html#construct-homepage-panels) hook, and each one is responsible for providing its own HTML rendering. In this way, the module providing the panel has full control over how it appears on the homepage.

Reviewer: ```suggestion
Working with objects that know how to render themselves as HTML using a template is a common pattern seen throughout the Wagtail admin. For example, the admin homepage is a view provided by the central `wagtail.admin` app, but brings together information panels sourced from various other modules of Wagtail, such as images and documents (potentially along with others provided by third-party packages). These panels are passed to the homepage via the [`construct_homepage_panels`](../reference/hooks.html#construct-homepage-panels) hook, and each one is responsible for providing its own HTML rendering. In this way, the module providing the panel has full control over how it appears on the homepage.
```

Author: The latter version here isn't strictly true - components don't necessarily use templates for their own rendering. (They might just return `mark_safe("<p>some static HTML</p>")` from the `render_html` method instead.) The message I'm aiming to get across here is that we're often faced with a page template, with a bunch of objects to render into it, and we _aren't_ particularly concerned with where those objects come from or how they arrive at their HTML.

```eval_rst
.. method:: render_html(self, parent_context=None)

Given a context dictionary from the calling template (which may be a :py:class:`Context <django.template.Context>` object or a plain ``dict`` of context variables), returns the string representation to be inserted into the template. This will be subject to Django's HTML escaping rules, so a return value consisting of HTML should typically be returned as a :py:mod:`SafeString <django.utils.safestring>` instance.

.. attribute:: media

A (possibly empty) :doc:`form media <django:topics/forms/media>` object defining JavaScript and CSS resources used by the component.

.. note::
   Any object implementing this API can be considered a valid component; it does not necessarily have to inherit from the ``Component`` class described below, and user code that works with components should not assume this (for example, it must not use ``isinstance`` to check whether a given value is a component).
```

```eval_rst
.. _creating_template_components:
```
## Creating components

> The simplest way to create a component is to define a subclass of `wagtail.admin.ui.components.Component` and specify a `template_name` attribute on it. The rendered template will then be used as the component's HTML representation:

Reviewer 2: If this is the simplest way, and Wagtail itself uses this way... Should we mention this is the _preferred_ way?

Author: Yes, I think we can say that! In #7409 the Cell and Header objects don't use the Component base class, because while they outwardly behave as components they actually delegate the HTML rendering to the Column object - and I wanted to avoid implying that was the "wrong" way to do things. But that's really a bit of a special case, and it's fair to describe the Component base class as the preferred way even if there are occasional exceptions.

```python
from wagtail.admin.ui.components import Component

class WelcomePanel(Component):
    template_name = 'my_app/panels/welcome.html'

my_welcome_panel = WelcomePanel()
```

`my_app/templates/my_app/panels/welcome.html`:

```html+django
<h1>Welcome to my app!</h1>
```

For simple cases that don't require a template, the `render_html` method can be overridden instead:

```python
from django.utils.html import format_html
from wagtail.admin.components import Component

class WelcomePanel(Component):
    def render_html(self, parent_context):
        return format_html("<h1>{}</h1>", "Welcome to my app!")
```

## Passing context to the template

The `get_context` method can be overridden to pass context variables to the template. As with `render_html`, this receives the context dictionary from the calling template:

```python
from wagtail.admin.ui.components import Component

class WelcomePanel(Component):
    template_name = 'my_app/panels/welcome.html'

>     def get_context(self, parent_context):
>         context = super().get_context(parent_context)
>         context['username'] = parent_context['request'].user.username
>         return context

Reviewer: ```suggestion
    def get_context_data(self, parent_context):
        context = super().get_context_data(parent_context)
        context['username'] = parent_context['request'].user.username
        return context
```

```

`my_app/templates/my_app/panels/welcome.html`:

```html+django
<h1>Welcome to my app, {{ username }}!</h1>
```

## Adding media definitions

Like Django form widgets, components can specify associated JavaScript and CSS resources using either an inner `Media` class or a dynamic `media` property:

```python
class WelcomePanel(Component):
    template_name = 'my_app/panels/welcome.html'

    class Media:
        css = {
            'all': ('my_app/css/welcome-panel.css',)
        }
```

> ## Using components on your own templates

Reviewer: ```suggestion
## Using components in your own templates
```

Author: Mmm... to me "on your own templates" feels like it's more closely pinpointing the action of putting a tag on a page - which is what this section is about - while "in your own templates" would cover the broader concepts of using components, which would include things like setting up a hook to collect components from outside code. That might just be a personal quirk though, and I'm not married to it :-)

Reviewer 2: `your own` -> `custom`?

Author: This section of the docs is aimed at developers building their own areas of the admin (e.g. as third-party packages), and it feels to me like that's beyond the point where it can reasonably be called "custom".
